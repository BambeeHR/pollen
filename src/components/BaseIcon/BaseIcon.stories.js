import BaseIcon, { Icons } from './BaseIcon.vue';
import '../../assets/tailwind.css';

export default {
  title: 'BaseIcon',
  component: BaseIcon,
};

export const Gallery = () => ({
  components: { BaseIcon },
  data() {
    return {
      icons: Object.values(Icons),
    };
  },
  template: `
    <div class="flex flex-wrap">
      <div v-for="icon in icons" :key="icon" :style="{ margin: '12px', textAlign: 'center' }">
        <BaseIcon :icon="icon" />
        <div :style="{ fontSize: '12px' }">{{ icon }}</div>
      </div>
    </div>
  `,
});
