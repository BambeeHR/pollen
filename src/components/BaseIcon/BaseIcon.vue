<template functional>
  <component
    :is="$options.getComponent(props.icon)"
    :class="[
      data.class,
      data.staticClass,
      { 'base-icon--custom': $options.isCustomIcon(props.icon) },
    ]"
    :icon="$options.getIcon(props.icon)"
    class="base-icon"
    v-bind="$options.getAttributes({ data, parent })"
    v-on="listeners"
  />
</template>

<script>
import camelCase from 'lodash/camelCase';
import snakeCase from 'lodash/snakeCase';
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome';
import { library } from '@fortawesome/fontawesome-svg-core';

import {
  faAnalytics as analytics,
  faBallotCheck as ballotCheckSolid,
  faClipboardCheck as clipboardCheckSolid,
  faCreditCard as creditCardSolid,
  faEar as ear,
  faTachometer as tachometerSolid,
  faUser as userSolid,
  faArrowLeft as arrowLeft,
  faArrowRight as arrowRight,
  faBan as ban,
  faBell as bell,
  faBuilding as building,
  faCabinetFiling as cabinetFilingSolid,
  faCalendar as calendar,
  faCheck as check,
  faCheckCircle as checkCircleSolid,
  faClock as clock,
  faCloudUploadAlt as upload,
  faCogs as cogsSolid,
  faExclamation as alert,
  faExclamationCircle as alertCircle,
  faExclamationTriangle as alertTriangle,
  faGift as giftSolid,
  faLock as lock,
  faPen as pen,
  faPhoneAlt as phone,
  faPrint as print,
  faQuestion as question,
  faQuestionCircle as questionCircle,
  faSignOutAlt as signOutSolid,
  faUmbrella as umbrellaSolid,
  faUsers as usersSolid,
  faTrash as trash,
  faVirus as virusSolid,
  faWrench as wrench,
} from '@fortawesome/pro-solid-svg-icons';

import {
  faCheckCircle as checkCircle,
  faFolder as folder,
  faSearch as search,
} from '@fortawesome/pro-regular-svg-icons';

import {
  faBallotCheck as ballotCheck,
  faCabinetFiling as cabinetFiling,
  faCircle as circle,
  faClipboardCheck as clipboardCheck,
  faCogs as cogs,
  faCreditCard as creditCard,
  faGift as gift,
  faPlus as plus,
  faSignOutAlt as signOut,
  faTimes as close,
  faTachometer as tachometer,
  faUmbrella as umbrella,
  faUser as user,
  faUsers as users,
  faVirus as virus,
} from '@fortawesome/pro-light-svg-icons';

import caretDown from '../../assets/icons/caret-down.svg';
import eyeClosed from '../../assets/icons/eye-closed.svg';
import eyeOpen from '../../assets/icons/eye-open.svg';

import getAttributes from '../../util/getAttributes';

const FAIcons = Object.freeze({
  alert,
  alertCircle,
  alertTriangle,
  analytics,
  arrowLeft,
  arrowRight,
  ballotCheck,
  ballotCheckSolid,
  ban,
  bell,
  building,
  cabinetFiling,
  cabinetFilingSolid,
  calendar,
  check,
  checkCircle,
  checkCircleSolid,
  circle,
  clipboardCheck,
  clipboardCheckSolid,
  clock,
  close,
  cogs,
  cogsSolid,
  creditCard,
  creditCardSolid,
  ear,
  folder,
  gift,
  giftSolid,
  lock,
  pen,
  phone,
  plus,
  print,
  question,
  questionCircle,
  search,
  signOut,
  signOutSolid,
  tachometer,
  tachometerSolid,
  trash,
  umbrella,
  umbrellaSolid,
  upload,
  user,
  userSolid,
  users,
  usersSolid,
  virus,
  virusSolid,
  wrench,
});

Object.values(FAIcons).forEach((icon) => {
  library.add(icon);
});

const CustomIcons = Object.freeze({
  caretDown,
  eyeClosed,
  eyeOpen,
});

export const Icons = Object.freeze(
  [...Object.keys(FAIcons), ...Object.keys(CustomIcons)].sort().reduce(
    (obj, val) => ({
      ...obj,
      [snakeCase(val).toUpperCase()]: val,
    }),
    {}
  )
);

export const isValidIcon = (value) =>
  Object.values(Icons).includes(camelCase(value));

const isCustomIcon = (iconName) => Object.keys(CustomIcons).includes(iconName);

const getComponent = (iconName) =>
  isCustomIcon(iconName) ? CustomIcons[iconName] : FontAwesomeIcon;

/**
 * `import { BaseIcon } from '@bambee/pollen';`
 *
 * An SVG icon component pulling from Bambee's predefined icon set. Icons are
 * designed to be styled according to the contextual size and font color, e.g.
 * if you want a red, 40px icon, make sure it or its containing element is
 * styled with `color: red; font-size: 40px;`.
 */
export default {
  props: {
    /** The name of the icon to use. */
    icon: {
      type: String,
      required: true,
      validator: isValidIcon,
    },
  },
  getAttributes,
  getComponent,
  getIcon: (icon) => FAIcons[camelCase(icon)] || null,
  isCustomIcon,
};
</script>

<style>
.base-icon--custom {
  height: 1em;
  width: 1em;
}
</style>
