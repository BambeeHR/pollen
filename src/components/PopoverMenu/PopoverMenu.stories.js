import { action } from '@storybook/addon-actions';
import BaseButton from '../BaseButton/BaseButton.vue';
import CounterBadge from '../CounterBadge/CounterBadge.vue';
import PopoverMenu from './PopoverMenu.vue';
import { Icons } from '../BaseIcon/BaseIcon.vue';

export default {
  title: 'PopoverMenu',
  component: PopoverMenu,
  parameters: {
    componentSubtitle: '',
  },
};

export const Default = () => ({
  components: { BaseButton, PopoverMenu },
  data() {
    return {
      items: [
        {
          label: 'Profile',
          handler: action('clicked Profile'),
        },
        {
          icon: Icons.CHECK,
          label: 'Settings',
          handler: action('clicked Settings'),
        },
        {
          icon: Icons.SIGN_OUT_SOLID,
          label: 'Sign Out',
          handler: action('clicked Sign Out'),
        },
      ],
    };
  },
  template: `
    <PopoverMenu 
      :items="items" 
    >
      <BaseButton tag="div">Toggle Popover</BaseButton>
    </PopoverMenu>
  `,
});

export const RightAligned = () => ({
  components: { BaseButton, PopoverMenu },
  data() {
    return {
      items: [
        {
          label: 'Profile',
          handler: action('clicked Profile'),
        },
        {
          icon: Icons.CHECK,
          label: 'Settings',
          handler: action('clicked Settings'),
        },
        {
          icon: Icons.SIGN_OUT_SOLID,
          label: 'Sign Out',
          handler: action('clicked Sign Out'),
        },
      ],
    };
  },
  template: `
    <PopoverMenu 
      alignment="right"
      :items="items" 
    >
      <BaseButton tag="div">Toggle Popover</BaseButton>
    </PopoverMenu>
  `,
});

export const DefinedWidth = () => ({
  components: { BaseButton, PopoverMenu },
  data() {
    return {
      items: [
        {
          label: 'Profile',
          handler: action('clicked Profile'),
        },
        {
          icon: Icons.CHECK,
          label: 'Settings',
          handler: action('clicked Settings'),
        },
        {
          icon: Icons.SIGN_OUT_SOLID,
          label: 'Sign Out',
          handler: action('clicked Sign Out'),
        },
      ],
    };
  },
  template: `
  <div class="relative">
    <PopoverMenu 
      width="300px"
      :items="items" 
    >
      <BaseButton tag="div">Toggle Popover</BaseButton>
    </PopoverMenu>
  </div>
  `,
});

export const InitiallyOpen = () => ({
  components: { BaseButton, PopoverMenu },
  data() {
    return {
      items: [
        {
          label: 'Profile',
          handler: action('clicked Profile'),
        },
        {
          icon: Icons.CHECK,
          label: 'Settings',
          handler: action('clicked Settings'),
        },
        {
          icon: Icons.SIGN_OUT_SOLID,
          label: 'Sign Out',
          handler: action('clicked Sign Out'),
        },
      ],
    };
  },
  template: `
    <PopoverMenu 
      :items="items" 
      initially-open
    >
      <BaseButton tag="div">Toggle Popover</BaseButton>
    </PopoverMenu>
  `,
});

export const WithSlotContent = () => ({
  components: { BaseButton, CounterBadge, PopoverMenu },
  data() {
    return {
      items: [
        {
          label: 'notifications',
        },
        {
          label: 'Profile',
          handler: action('clicked Profile'),
        },
        {
          icon: Icons.CHECK,
          label: 'Settings',
          handler: action('clicked Settings'),
        },
        {
          icon: Icons.SIGN_OUT_SOLID,
          label: 'Sign Out',
          handler: action('clicked Sign Out'),
        },
      ],
    };
  },
  template: `
  <PopoverMenu
    :items="items"
  >
    <BaseButton tag="div">Toggle Popover</BaseButton>
    <template v-slot:notifications>
      <div class="flex items-center px-4 py-2">
        Notifications 
        <CounterBadge :value="47" class="ml-2" />
      </div>
    </template>
  </PopoverMenu>
  `,
});
