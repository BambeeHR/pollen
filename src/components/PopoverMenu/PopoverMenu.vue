<template>
  <div class="popover-menu">
    <button
      class="popover-menu__trigger"
      type="button"
      @click.stop="isActive = !isActive"
    >
      <slot />
    </button>
    <div
      v-if="isActive"
      class="popover-menu__items"
      :class="`popover-menu__items--${alignment}`"
      :style="{ width }"
    >
      <template v-for="{ handler = () => {}, icon, label } in items">
        <slot :name="label">
          <button :key="label" class="popover-menu__item" @click="handler">
            <BaseIcon v-if="icon" class="popover-menu__icon" :icon="icon" />
            <span>{{ label }}</span>
          </button>
        </slot>
      </template>
    </div>
  </div>
</template>

<script>
import BaseIcon from '../BaseIcon/BaseIcon.vue';

const Alignments = Object.freeze({
  LEFT: 'left',
  RIGHT: 'right',
});
export default {
  components: { BaseIcon },
  props: {
    /**
     * Whether the menu should align with the `left` or `right` of the trigger element.
     */
    alignment: {
      type: String,
      default: Alignments.LEFT,
    },
    /**
     * If true, the PopoverMenu will appear open when first rendered.
     */
    initiallyOpen: {
      type: Boolean,
      default: false,
    },
    /**
     * An array of items to display. These should be objects with a `label`,
     * `handler`, and `icon` (optional) keys. For custom content, slots can be
     * used, in which case only a `label` is required that will serve as the
     * slot name.
     */
    items: {
      type: Array,
      default: () => [],
    },
    /**
     * The width of the menu itself.
     */
    width: {
      type: String,
      default: 'auto',
    },
  },
  data() {
    return { isActive: this.initiallyOpen };
  },
  watch: {
    initiallyOpen(newValue) {
      this.isActive = newValue;
    },
    isActive(newValue) {
      if (newValue) {
        this.addListeners();
      } else {
        this.removeListeners();
      }
    },
  },
  mounted() {
    if (this.initiallyOpen) {
      this.addListeners();
    }
  },
  beforeDestroy() {
    this.removeListeners();
  },
  methods: {
    addListeners() {
      document.addEventListener('keyup', this.handleKeyup);
      document.body.addEventListener('click', this.close);
    },
    close() {
      this.isActive = false;
    },
    handleKeyup(e) {
      if (e.keyCode !== 27) {
        return;
      }
      this.close();
    },
    removeListeners() {
      document.removeEventListener('keyup', this.handleKeyup);
      document.body.removeEventListener('click', this.close);
    },
  },
};
</script>

<style scoped>
.popover-menu {
  @apply inline-block relative;
}

.popover-menu__trigger {
  @apply bg-transparent border-none p-0;
}

.popover-menu__items {
  @apply absolute
    bg-white 
    bottom-0
    py-2 
    rounded-md 
    transform
    translate-y-full
    shadow-3;
}

.popover-menu__items--left {
  @apply left-0;
}

.popover-menu__items--right {
  @apply right-0;
}

.popover-menu__item {
  @apply bg-white 
    border-none 
    cursor-pointer 
    flex 
    font-body-text 
    font-medium 
    items-center 
    px-4 
    py-2 
    text-left 
    truncate
    w-full;
}

.popover-menu__item:hover,
.popover-menu__item:focus {
  @apply bg-gray-7;
}

.popover-menu__item:active {
  @apply bg-gray-6;
}

.popover-menu__icon {
  @apply flex-shrink-0 mr-2;
}
</style>
