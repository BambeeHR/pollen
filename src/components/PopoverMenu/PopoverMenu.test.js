import { mount } from '@vue/test-utils';
import PopoverMenu from './PopoverMenu.vue';
import { Icons } from '../BaseIcon/BaseIcon.vue';

describe('PopoverMenu', () => {
  let items;
  let handler;
  beforeEach(() => {
    handler = jest.fn();
    items = [
      {
        label: 'Profile',
        handler,
      },
      {
        icon: Icons.CHECK,
        label: 'Settings',
        handler,
      },
      {
        icon: Icons.SIGN_OUT_SOLID,
        label: 'Sign Out',
        handler,
      },
    ];
  });

  test('renders correctly', () => {
    const wrapper = mount(PopoverMenu, {
      propsData: { initiallyOpen: true, items },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  test('renders empty correctly', () => {
    const wrapper = mount(PopoverMenu, { propsData: { initiallyOpen: true } });
    expect(wrapper.element).toMatchSnapshot();
  });

  test('makes the menu visible when the trigger is clicked', async () => {
    const wrapper = mount(PopoverMenu, {
      propsData: { items },
    });
    expect(wrapper.find('.popover-menu__items').exists()).toBe(false);
    await wrapper.find('.popover-menu__trigger').trigger('click');
    expect(wrapper.find('.popover-menu__items').exists()).toBe(true);
    await wrapper.find('.popover-menu__trigger').trigger('click');
    expect(wrapper.find('.popover-menu__items').exists()).toBe(false);
  });

  test('executes the nav item handler on click', () => {
    const wrapper = mount(PopoverMenu, {
      propsData: { initiallyOpen: true, items },
    });
    wrapper.find('.popover-menu__items button').trigger('click');
    expect(handler).toHaveBeenCalled();
  });

  test('updating `initiallyOpen` updates the popover state', async () => {
    const wrapper = mount(PopoverMenu);
    expect(wrapper.vm.isActive).toBe(false);
    await wrapper.setProps({ initiallyOpen: true });
    expect(wrapper.vm.isActive).toBe(true);
  });

  test('closes the menu when escape is pressed', async () => {
    const elem = document.createElement('div');
    if (document.body) {
      document.body.appendChild(elem);
    }
    const wrapper = mount(PopoverMenu, {
      attachTo: elem,
      propsData: { initiallyOpen: true },
    });
    expect(wrapper.vm.isActive).toBe(true);
    await wrapper.trigger('keyup.enter');
    expect(wrapper.vm.isActive).toBe(true);
    await wrapper.trigger('keyup.esc');
    expect(wrapper.vm.isActive).toBe(false);
  });

  test('closes the menu when the document is clicked', async () => {
    const wrapper = mount(PopoverMenu, {
      propsData: { initiallyOpen: true },
    });
    expect(wrapper.vm.isActive).toBe(true);
    await document.body.click();
    expect(wrapper.vm.isActive).toBe(false);
  });

  test('adds and removes event listeners', () => {
    const { methods } = PopoverMenu;
    const addListenerSpy = jest.spyOn(methods, 'addListeners');
    const removeListenerSpy = jest.spyOn(methods, 'removeListeners');
    const wrapper = mount(PopoverMenu, {
      propsData: { initiallyOpen: true },
    });
    wrapper.destroy();
    expect(addListenerSpy).toHaveBeenCalled();
    expect(removeListenerSpy).toHaveBeenCalled();
  });
});
