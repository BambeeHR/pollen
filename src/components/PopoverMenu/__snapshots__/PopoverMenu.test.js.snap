// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopoverMenu renders correctly 1`] = `
<div
  class="popover-menu"
>
  <button
    class="popover-menu__trigger"
    type="button"
  />
   
  <div
    class="popover-menu__items popover-menu__items--left"
    style="width: auto;"
  >
    <button
      class="popover-menu__item"
    >
      <!---->
       
      <span>
        Profile
      </span>
    </button>
    <button
      class="popover-menu__item"
    >
      <svg
        aria-hidden="true"
        class="base-icon svg-inline--fa fa-check fa-w-16 popover-menu__icon"
        data-icon="check"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class=""
          d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
          fill="currentColor"
        />
      </svg>
       
      <span>
        Settings
      </span>
    </button>
    <button
      class="popover-menu__item"
    >
      <svg
        aria-hidden="true"
        class="base-icon svg-inline--fa fa-sign-out-alt fa-w-16 popover-menu__icon"
        data-icon="sign-out-alt"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class=""
          d="M497 273L329 441c-15 15-41 4.5-41-17v-96H152c-13.3 0-24-10.7-24-24v-96c0-13.3 10.7-24 24-24h136V88c0-21.4 25.9-32 41-17l168 168c9.3 9.4 9.3 24.6 0 34zM192 436v-40c0-6.6-5.4-12-12-12H96c-17.7 0-32-14.3-32-32V160c0-17.7 14.3-32 32-32h84c6.6 0 12-5.4 12-12V76c0-6.6-5.4-12-12-12H96c-53 0-96 43-96 96v192c0 53 43 96 96 96h84c6.6 0 12-5.4 12-12z"
          fill="currentColor"
        />
      </svg>
       
      <span>
        Sign Out
      </span>
    </button>
  </div>
</div>
`;

exports[`PopoverMenu renders empty correctly 1`] = `
<div
  class="popover-menu"
>
  <button
    class="popover-menu__trigger"
    type="button"
  />
   
  <div
    class="popover-menu__items popover-menu__items--left"
    style="width: auto;"
  />
</div>
`;
